### PR TITLE
Support range, cidrs in grepip

### PR DESCRIPTION
**Pull request by talhahwahla**
Ref: https://github.com/ipinfo/cli/pull/185#issuecomment-1809975092

### CODE REVIEW

File: ipinfo/cmd_grepip.go [hunk 1]
```diff
@@ -13,6 +13,12 @@ var completionsGrepIP = &complete.Command{
 	Flags: map[string]complete.Predictor{
 		"-o":                 predict.Nothing,
 		"--only-matching":    predict.Nothing,
+		"-c":                 predict.Nothing,
+		"--include-cidrs":    predict.Nothing,
+		"-r":                 predict.Nothing,
```
Comment (UmanShahzad) [inline ipinfo/cmd_grepip.go:18]: Let's not use a short flag for this - we tried to copy some common flag naming conventions from normal `grep` (hence `-o`), and `-r` is too easily associated with `recursive lookup`.

File: ipinfo/cmd_grepip.go [hunk 2]
```diff
@@ -13,6 +13,12 @@ var completionsGrepIP = &complete.Command{
 	Flags: map[string]complete.Predictor{
 		"-o":                 predict.Nothing,
 		"--only-matching":    predict.Nothing,
+		"-c":                 predict.Nothing,
```
Comment (UmanShahzad) [inline ipinfo/cmd_grepip.go:16]: Same here, this is too easily associated with `count` in `grep`

File: ipinfo/cmd_grepip.go [hunk 3]
```diff
@@ -35,6 +41,14 @@ Options:
   General:
     --only-matching, -o
       print only matched IP in result line, excluding surrounding content.
+    --include-cidrs, -c
+      prints the CIDRs too.
+    --include-ranges, -r
```
Comment (UmanShahzad) [inline ipinfo/cmd_grepip.go:46]: Rm `-r`

File: ipinfo/cmd_grepip.go [hunk 4]
```diff
@@ -35,6 +41,14 @@ Options:
   General:
     --only-matching, -o
       print only matched IP in result line, excluding surrounding content.
+    --include-cidrs, -c
```
Comment (UmanShahzad) [inline ipinfo/cmd_grepip.go:44]: Rm `-c`

File: lib/cmd_grepip.go [hunk 5]
```diff
@@ -37,6 +41,26 @@ func (f *CmdGrepIPFlags) Init() {
 		"only-matching", "o", false,
 		"print only matched IPs in result line.",
 	)
+	pflag.BoolVarP(
+		&f.IncludeCIDRs,
+		"include-cidrs", "c", false,
```
Comment (UmanShahzad) [inline lib/cmd_grepip.go:46]: Rm short

File: lib/cmd_grepip.go [hunk 6]
```diff
@@ -37,6 +41,26 @@ func (f *CmdGrepIPFlags) Init() {
...
+		"print cidrs too.",
+	)
+	pflag.BoolVarP(
+		&f.IncludeRanges,
+		"include-ranges", "r", false,
```
Comment (UmanShahzad) [inline lib/cmd_grepip.go:51]: Rm short

File: lib/init.go [hunk 7]
```diff
@@ -5,7 +5,32 @@ import "regexp"
 func init() {
```
Comment (UmanShahzad) [inline lib/init.go:5]: Do the new regexes below add any significant startup time to the command? With so many regexes compiled at startup, wanna make sure the command still runs super fast without overheads, as these regexes would be getting compiled for every single invocation of the CLI no matter what command.

File: lib/init.go [hunk 8]
```diff
@@ -5,7 +5,32 @@ import "regexp"
...
 	bogonIP4List = GetBogonRange4()
 	bogonIP6List = GetBogonRange6()
+
 	ipV4Regex = regexp.MustCompilePOSIX(IPv4RegexPattern)
 	ipV6Regex = regexp.MustCompilePOSIX(IPv6RegexPattern)
 	ipRegex = regexp.MustCompilePOSIX(IPv4RegexPattern + "|" + IPv6RegexPattern)
+
+	v4IpCidrRegex = regexp.MustCompilePOSIX(IPv4RegexPattern + "|" + IPv4CIDRRegexPattern)
```
Comment (UmanShahzad) [inline lib/init.go:13]: A comment against each grouped code is warranted now

File: lib/cmd_grepip.go [hunk 9]
```diff
@@ -121,10 +145,46 @@ func CmdGrepIP(
 
 	// prepare regexp
 	var rexp *regexp.Regexp
-	if ipv == 4 {
+	if ipv == 4 && f.CIDRsOnly && f.RangesOnly {
```
Comment (UmanShahzad) [inline lib/cmd_grepip.go:148]: Can we find a grouping that makes this more readable? If nesting `if`s helps then let's try that. Needs a judgement

File: lib/regex.go [hunk 10]
```diff
@@ -6,5 +6,35 @@ var ipV4Regex *regexp.Regexp
...
+
+var v4RangeRegex *regexp.Regexp
+var v6RangeRegex *regexp.Regexp
+var rangeRegex *regexp.Regexp
+
+var v4SubnetRegex *regexp.Regexp
+var v6SubnetRegex *regexp.Regexp
+var subnetRegex *regexp.Regexp
+
+const octet = `(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)`
```
Comment (UmanShahzad) [inline lib/regex.go:33]: It's just a `v4Octet`

File: lib/regex.go [hunk 11]
```diff
@@ -6,5 +6,35 @@ var ipV4Regex *regexp.Regexp
...
+const IPv4RegexPattern = `(` + octet + `\.){3}` + octet
+const IPv4RangeRegexPattern = IPv4RegexPattern + `[:space:]*[-,][:space:]*` + IPv4RegexPattern
+const IPv4CIDRRegexPattern = IPv4RegexPattern + `\/([1-2]?[0-9]|3[0-2])`
+
 const IPv6RegexPattern = `(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))`
+const IPv6RangeRegexPattern = IPv6RegexPattern + `[:space:]*[-,][:space:]*` + IPv6RegexPattern
```
Comment (UmanShahzad) [inline lib/regex.go:39]: Not sure about the infinitely allowed spaces. Our common definition of ranges has always been without any space separation allowed.

File: lib/regex.go [hunk 12]
```diff
@@ -6,5 +6,35 @@ var ipV4Regex *regexp.Regexp
...
+var v6RangeRegex *regexp.Regexp
+var rangeRegex *regexp.Regexp
+
+var v4SubnetRegex *regexp.Regexp
+var v6SubnetRegex *regexp.Regexp
+var subnetRegex *regexp.Regexp
+
+const octet = `(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)`
+const IPv4RegexPattern = `(` + octet + `\.){3}` + octet
+const IPv4RangeRegexPattern = IPv4RegexPattern + `[:space:]*[-,][:space:]*` + IPv4RegexPattern
```
Comment (UmanShahzad) [inline lib/regex.go:35]: Same

File: lib/init.go (hunk 7 above)
Comment (talhahwahla) [reply lib/init.go:5]:
```
$ time ipinfo grepip list.txt 
...
real	0m0.010s
user	0m0.000s
sys	0m0.011s
```

```
$ go build
$ time ./ipinfo grepip list.txt 
...
real	0m0.023s
user	0m0.009shas
sys	0m0.023s
```
execution time has increased now, difference in practice doesn't appear to be as significant, I guess

Comment (UmanShahzad) [reply lib/init.go:5]: Went up 13ms, so should be fine for now. We  can optimize if it gets worse

**Review by UmanShahzad — APPROVED** [review]

**Pull request merged**